### PR TITLE
NMS-13100: Calculate aggregations on-top of each other

### PR DESCRIPTION
**Pull request by swachter**
https://issues.opennms.org/browse/NMS-13100

3 aggregations, namely exporter/interface/tos/(app, conv, host) are calculated. By regrouping, all other aggregations arre derived therefrom.

### CODE REVIEW

File: main/src/main/java/org/opennms/nephron/Pipeline.java [hunk 1]
```diff
@@ -673,4 +591,113 @@ public void processElement(ProcessContext c, IntervalWindow window) {
...
+
+    public static class SumAndTopK {
+        public final PCollection<KV<CompoundKey, Aggregate>> sum;
+        public final PCollection<FlowSummaryData> topK;
+
+        public SumAndTopK(PCollection<KV<CompoundKey, Aggregate>> sum, PCollection<FlowSummaryData> topK) {
+            this.sum = sum;
+            this.topK = topK;
+        }
+    }
```
Comment (fooker) [inline main/src/main/java/org/opennms/nephron/Pipeline.java:691]: Can these two classes be unified?

**Review by fooker — CHANGES_REQUESTED** [review]
Maybe the handling of the `CompoundKey` handling regarding "with TOS" and "without TOS" can be further simplified: The order of elements in the key does not really matter, right? If we put TOS last in the key, this would simply result in some kind of `parent()` operation which drops the last element in the compound key. This also applies for the compound key type.

File: main/src/main/java/org/opennms/nephron/Ref.java [hunk 2]
```diff
@@ -42,6 +42,12 @@
  */
 abstract class Ref {
 
+    /**
+     * Returns a string representation of this key part or {@code null} in case this key part is ignored
+     * during aggregation.
+     *
+     * @return
+     */
```
[the remark this answers is not in the document]
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/Ref.java:50]: Good catch. Can no more be `null`. Was a leftover from the ecn related keying logic. I change that comment and remove the null-ness filter below.

File: main/src/main/java/org/opennms/nephron/Pipeline.java [hunk 3]
```diff
@@ -652,8 +570,8 @@ private Aggregate aggregatize(final IntervalWindow window, final FlowDocument fl
             // Track
             flowsInWindow.inc();
             return Direction.INGRESS.equals(flow.getDirection()) ?
-                   new Aggregate(bytes, 0, hostname, flow.getEcn().getValue()):
-                   new Aggregate(0, bytes, hostname, flow.getEcn().getValue());
+                   new Aggregate(bytes, 0, hostname, flow.hasEcn() ? flow.getEcn().getValue() : null):
+                   new Aggregate(0, bytes, hostname, flow.hasEcn() ? flow.getEcn().getValue() : null);
```
[the remark this answers is not in the document]
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/Pipeline.java:574]: Yes. (Caused by not relying on `tos` as it originally did.)

File: main/src/main/java/org/opennms/nephron/CompoundKey.java [hunk 4]
```diff
@@ -86,8 +87,30 @@ public CompoundKey getOuterKey() {
...
+        if (j != projectedType.getParts().length) {
+            // the parts of this key could do not match the parts of the projected type
+            throw new RuntimeException("key of type " + type + " can not be project into key of type " + projectedType);
+        }
+        return new CompoundKey(projectedType, l);
+    }
+
     public String groupedByKey() {
-        return refs.stream().map(Ref::idAsString).collect(Collectors.joining("-"));
+        return refs.stream().map(Ref::idAsString).filter(s -> s != null).collect(Collectors.joining("-"));
```
[the remark this answers is not in the document]
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/CompoundKey.java:113]: removed filter

File: main/src/main/java/org/opennms/nephron/Pipeline.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fooker) [reply main/src/main/java/org/opennms/nephron/Pipeline.java:574]: C'mon. Give it a separate commit and an appropriate commit message. And maybe link this to NMS-13099 (by doing so in the commit message=

**Comment by fooker** [comment]
Overall style: The commit message lacks the issue number according to our agreed styles. Same was true for the PR title. And it would be nice to have a sentence or two in the PR description explaining what this change does (in contrast to the issue, which explains the problem, not the solution).

File: main/src/main/java/org/opennms/nephron/Pipeline.java (hunk 1 above)
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/Pipeline.java:691]: Yes, they could be unified. However, I think the code is more readable if separate classes are used. I would prefer to leave it as is.

File: main/src/main/java/org/opennms/nephron/Pipeline.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/Pipeline.java:574]: I could add a commit to the `nms-12945` branch which does the `.hasEcn()` and rebase this branch again. on `nms-12945`. Estimated time: 1h. Is this what you want?

**Comment by swachter** [comment]
> Maybe the handling of the `CompoundKey` handling regarding "with TOS" and "without TOS" can be further simplified: The order of elements in the key does not really matter, right? If we put TOS last in the key, this would simply result in some kind of `parent()` operation which drops the last element in the compound key. This also applies for the compound key type.

Alternatively, we could leave the `CompoundKeyType` class unchanged and replace the currently used `CompoundKey.project()` method by a method that just drops the second last key part.

File: main/src/main/java/org/opennms/nephron/Pipeline.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fooker) [reply main/src/main/java/org/opennms/nephron/Pipeline.java:572]: Just make it two commits and describe in the commit message what the reason for this is. It should take not more than 10 minutes.

**Comment by fooker** [comment]
> > Maybe the handling of the `CompoundKey` handling regarding "with TOS" and "without TOS" can be further simplified: The order of elements in the key does not really matter, right? If we put TOS last in the key, this would simply result in some kind of `parent()` operation which drops the last element in the compound key. This also applies for the compound key type.
> 
> Alternatively, we could leave the `CompoundKeyType` class unchanged and replace the currently used `CompoundKey.project()` method by a method that just drops the second last key part.

If you don't like to change the structure of the keys, the current implementation is fine.

File: main/src/main/java/org/opennms/nephron/Pipeline.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/Pipeline.java:572]: done

**Review by fooker — APPROVED** [review]

**Pull request merged**